[clang][CodeGen] Allow memcpy replace with trivial auto var init

antoniofrighetto opened the pull request:
When emitting the storage (or memory copy operations) for constant initializers, the decision whether to split a constant structure or array store into a sequence of field stores or to use `memcpy` is based upon the optimization level and the size of the initializer. In https://github.com/llvm/llvm-project/commit/afe8b93ffdfef5d8879e1894b9d7dda40dee2b8d, we extended this by allowing constants to be split when the array (or struct) type does not match the type of data the address to the object (constant) is expected to contain. This may happen when `emitStoresForConstant` is called by `EmitAutoVarInit`, as the element type of the address gets shrunk. When this occurs, let the initializer be split into a bunch of stores only under `-ftrivial-auto-var-init=pattern`.

Fixes regression: https://github.com/llvm/llvm-project/issues/84178.

nikic commented:
Can you please provide more information on what the regression was / how this fixes it?

antoniofrighetto commented:
I'm not exactly sure if this is meant to occur with `TrivialAutoVarInitKind::Uninitialized` too, but this is preventing `memcpy`s from being emitted, in favour of stores for each byte; even if the alloca [initializer is not small](https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGDecl.cpp#L1244) (this happens seemingly regardless of a recurring pattern in the string, as I think it should occur with `-ftrivial-auto-var-init`). Thus memcpys from string literals get expanded into single stores. If we want to keep the patch, I think we should put some upper bound on the `NumElements` of the alloca, although my feeling is that we should not do this kind of optimizations at this level, as it prevents other optimizations later in the middle-end to take place.

antoniofrighetto commented:
I think manually checking if `TrivialAutoVarInit` is set (to `Pattern`?) may be a better fix, as there was only [one codepath](https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGDecl.cpp#L1972-L1974) in which we were not checking this.

antoniofrighetto committed:
[clang][CodeGen] Allow `memcpy` replace with trivial auto var init

When emitting the storage (or memory copy operations) for constant
initializers, the decision whether to split a constant structure or
array store into a sequence of field stores or to use `memcpy` is
based upon the optimization level and the size of the initializer.
In afe8b93ffdfef5d8879e1894b9d7dda40dee2b8d, we extended this by
allowing constants to be split when the array (or struct) type does
not match the type of data the address to the object (constant) is
expected to contain. This may happen when `emitStoresForConstant` is
called by `EmitAutoVarInit`, as the element type of the address gets
shrunk. When this occurs, let the initializer be split into a bunch
of stores only under `-ftrivial-auto-var-init=pattern`.

Fixes: https://github.com/llvm/llvm-project/issues/84178.

antoniofrighetto commented:
Fixed `auto-var-init.cpp` test failures, believe now it should be aligned with the original intent.

satmandu commented:
Will there be a backport for 18.1.x? (The current patch doesn't apply cleanly to 18.1.0.)

AaronBallman commented:
You should update the patch summary since this is no longer a revert.

nikic commented:
This change also "fixes" https://github.com/llvm/llvm-project/issues/78034 in the sense that we return to the clang 17 status quo of the emitted IR being wrong but mostly working out in practice.

efriedma-quic reviewed (commented):
This is clearly not what we want long-term.  Either clang or LLVM should have better heuristics based on the actual values being stored; the syntax used at the C level shouldn't affect codegen.

That said, reverting the change with a carveout for the trivial var init is probably reasonable as a temporary step, if we want something we can cherry-pick for 18.x.

Please add a better comment explaining the history here; the current comment isn't really sufficient to explain why we're using heuristics which don't really make sense.

antoniofrighetto commented:
@efriedma-quic, comment updated, thanks.

nikic commented:
@antoniofrighetto You need to update the PR description, not the commit message.

antoniofrighetto commented:
@nikic, updated PR description as well, thanks.

AaronBallman reviewed (approved):
I think the changes here make sense (keeping in mind what @efriedma-quic says about long-term goals vs 18.x practicality). Eli, @serge-sans-paille, are you okay with the changes as well? (Please don't land on my approval without waiting a bit for Eli and Serge.)

If so, I'd like to get this backported into 18.x as well because of the issues we're seeing in the wild. Any disagreement?

efriedma-quic reviewed (approved):
LGTM

serge-sans-paille commented:
LGTM too. It's a bit sad that we couldn't make the optimization part of LLVM better instead of generating ad-hoc pattern, but as I couldn't find the time to work more on this, I'm not going to complain!

tstellar commented:
Do we need to backport this to release/18.x?

antoniofrighetto commented:
@tstellar Yes, thanks (I opened a new PR for that: https://github.com/llvm/llvm-project/pull/86106).
